Bootstrap ruff

kwryankrattiger opened the pull request:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

Add ruff to dev bootstrap and style tools. Ruff is added as a linting option, but is not part of the default/required linting tools.

CC: @adamjstewart @tgamblin

kwryankrattiger committed:
Add Ruff linter to dev bootstrap

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Add ruff to linting tool options

ruff is not officially used by Spack. This is currently being added
as an experimental option.

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

adamjstewart commented on lib/spack/spack/bootstrap/environment.py:
```diff
@@ -44,6 +44,7 @@ def spack_dev_requirements(cls) -> List[str]:
             black_root_spec(),
             flake8_root_spec(),
             pytest_root_spec(),
+            pyruff_root_spec(),
```
Why pyruff instead of ruff? All the other formatters don't have a py prefix.

kwryankrattiger replied:
Because I am bad at reading mostly :p I'll update that once I figure out why the bootstrap test is angry.

adamjstewart commented on lib/spack/spack/bootstrap/environment.py:
```diff
@@ -146,6 +147,11 @@ def pytest_root_spec() -> str:
 
+def pyruff_root_spec() -> str:
+    """Return the root spec used to bootstrap flake8"""
+    return _root_spec("py-ruff@0.8.0:")
```
Why 0.8?

kwryankrattiger replied:
It is the version that uses Python 3.9 by default as the target python version.

Eventually we would probably want to pin the version to avoid breaking linting changes. I didn't here since this is still experimental.

adamjstewart replied:
Why would we want Python 3.9?

kwryankrattiger replied:
Because 3.8 is EOL...

adamjstewart replied:
Spack still supports Python 3.6 unfortunately...

adamjstewart replied:
I would actually pick 0.9+, 0.8 and 0.9 have different styles and will reformat each other, they aren't backwards compatible like 0.7 and 0.8 were.

tgamblin commented:
LGTM but needs to actually concretize and bootstrap :)

kwryankrattiger commented:
So ruff requires python 3.7 to bootstrap.

```
py-ruff -> py-setuptools-rust@1.4: -> py-setuptools@62: -> python@3.7:
```

@tgamblin @alalazo Is this a requirement we can bump now for Spack v1? Or should ruff have another option to bootstrap that is excluded.

kwryankrattiger commented:
Verified locally that `py-ruff@0.9:` is bootstrapped when a newer python is available.

adamjstewart commented:
I'm pro dropping support for EOL Python versions, but this is a long-standing point of contention. Worth discussing in the TSC meeting tomorrow.

tgamblin commented:
> I'm pro dropping support for EOL Python versions, but this is a long-standing point of contention. Worth discussing in the TSC meeting tomorrow.

We don't have to drop support for Spack -- just for dev tools.  So basically:
* the bootstrap workflow shouldn't run on the runner we keep around to ensure that our tests work on RHEL8.  It's ok if it doesn't work there.
* `spack bootstrap --dev` can fail there.

We've talked about this a bit for `mypy` as well -- it is very hard to have good dev tooling *on* Python 3.6.  However, I suspect nobody is developing Spack itself on Python 3.6.  I suspect many people are *running* it on 3.6.

kwryankrattiger commented:
Before I invoke more churn on this (it is passing now) do we want to disable bootstrap dev completely for python 3.6 and dump a nice message or just exclude bootstrapping ruff for that version of python?

tgamblin commented:
I think just excluding ruff is fine for now.

adamjstewart commented:
I tried disabling dev tools on Python 3.6 one time but we use it to bootstrap pytest so I couldn't get it to pass without also removing 3.6 tests.

tgamblin commented:
@kwryankrattiger are you able to bootstrap locally w/this?

I get this error:

```console
> spack -t bootstrap now --dev
Traceback (most recent call last):
  File "/Users/gamblin2/Workspace/src/spack/bin/spack", line 50, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack_installable/main.py", line 41, in main
    sys.exit(spack.main.main(argv))
             ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/main.py", line 1038, in main
    return _main(argv)
           ^^^^^^^^^^^
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/main.py", line 991, in _main
    return finish_parse_and_run(parser, cmd_name, args, env_format_error)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/main.py", line 1021, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/main.py", line 589, in _invoke_command
    return_val = command(parser, args)
                 ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/cmd/bootstrap.py", line 463, in bootstrap
    callbacks[args.subcommand](args)
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/cmd/bootstrap.py", line 447, in _now
    spack.bootstrap.ensure_environment_dependencies()
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/bootstrap/environment.py", line 164, in ensure_environment_dependencies
    env.load()
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/bootstrap/environment.py", line 109, in load
    sys.path.extend(str(p) for p in self.python_dirs())
                                    ^^^^^^^^^^^^^^^^^^
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/bootstrap/environment.py", line 80, in python_dirs
    return {self.view_root().joinpath(p) for p in (python.platlib, python.purelib)}
                                                   ^^^^^^^^^^^^^^
  File "/Users/gamblin2/Workspace/src/spack/var/spack/repos/builtin/packages/python-venv/package.py", line 65, in platlib
    relative_platlib = os.path.relpath(self._get_path("platlib"), self.prefix)
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/gamblin2/Workspace/src/spack/var/spack/repos/builtin/packages/python-venv/package.py", line 58, in _get_path
    return self.command(
           ^^^^^^^^^^^^^
TypeError: 'NoneType' object is not callable
```

I think this may be a more general issue w/bootstrapping -- it's failing to find the python command, which I *believe* might have to do with how I set up my environment.

tgamblin commented:
Also AFAICT the coverage number is wrong - it looks like the patch has close to 100% coverage, so we won't worry about that.

kwryankrattiger commented:
I was able to bootstrap locally before the rebase, but I did not retry after the rebase. I presumed the CI bootstrap check would fail if there was some other critical issue beyond selecting the right packages to bootstrap.

I did not run into the error you posted here.

kwryankrattiger commented:
Ran this locally today an no errors on the latest version.

kwryankrattiger commented:
@tgamblin any luck getting your local bootstrap to work?

kwryankrattiger committed:
Update completions

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

adamjstewart commented on lib/spack/spack/bootstrap/environment.py:
```diff
@@ -145,6 +150,11 @@ def pytest_root_spec() -> str:
 
 
+def ruff_root_spec() -> str:
+    """Return the root spec used to bootstrap flake8"""
```
```suggestion
    """Return the root spec used to bootstrap ruff"""
```

adamjstewart commented on lib/spack/spack/bootstrap/environment.py:
```diff
@@ -145,6 +150,11 @@ def pytest_root_spec() -> str:
 
+def ruff_root_spec() -> str:
+    """Return the root spec used to bootstrap flake8"""
+    return _root_spec("py-ruff@0.9.0:")
```
We may want to pin to a specific version and update once per year. 0.9.0 introduced the 2025 style guide, but they're already up to 0.14.3.

kwryankrattiger replied:
should we pin 0.14.3? I think @tgamblin did run ruff at one point, not sure which version he used though.

adamjstewart replied:
He didn't merge the changes, so that doesn't really matter. Latest version in Spack is 0.13.0 but I can update if we want. It might not be crucial, they've been relatively stable since 0.9. Once the 2026 style updates come, we can reconsider.

kwryankrattiger replied:
ok, I will pin to 0.13.0 then and we can reevaluate later.

kwryankrattiger committed:
Pin ruff to 0.13.0

This is the current last version available in Spack. Pinning to avoid
formatting churn. To be reevaluated in 2026 with new standards.

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Error when creating bootstrap environment with incompatible python

Spack requires python>=3.6
Spack Dev requries python>=3.10

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

adamjstewart commented:
When can we drop 3.6?
